File: dqops/src/main/java/com/dqops/rest/models/comparison/TableComparisonGroupingColumnPairModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.rest.models.comparison;

import com.dqops.metadata.comparisons.TableComparisonGroupingColumnsPairSpec;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

/**
 * Model that identifies a pair of column names used for grouping the data on both the compared table and the reference table. The groups are then matched (joined) by DQOps to compare aggregated results.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "TableComparisonGroupingColumnPairModel", description = "Model that identifies a pair of column names used for grouping the data on both the compared table and the reference table. " +
        "The groups are then matched (joined) by DQOps to compare aggregated results.")
@Data
public class TableComparisonGroupingColumnPairModel {
    /**
     * The name of the column on the compared table (the parent table) that is used in the GROUP BY clause to group rows before compared aggregates (row counts, sums, etc.) are calculated.
     * This column is also used to join (match) results to the reference table.
     */
    @JsonPropertyDescription("The name of the column on the compared table (the parent table) that is used in the GROUP BY clause to group rows before compared " +
            "aggregates (row counts, sums, etc.) are calculated. This column is also used to join (match) results to the reference table.")
    private String comparedTableColumnName;

    /**
     * The name of the column on the reference table (the source of truth) that is used in the GROUP BY clause to group rows before compared " +
     * "aggregates (row counts, sums, etc.) are calculated. This column is also used to join (match) results to the compared table.
     */
    @JsonPropertyDescription("The name of the column on the reference table (the source of truth) that is used in the GROUP BY clause to group rows before compared " +
            "aggregates (row counts, sums, etc.) are calculated. This column is also used to join (match) results to the compared table.")
    private String referenceTableColumnName;

    /**
     * Creates the compared columns column model from the specification.
     * @param columnsPairSpec Source compared columns pair specification.
     * @return Model for the object.
     */
    public static TableComparisonGroupingColumnPairModel fromColumnPairSpec(TableComparisonGroupingColumnsPairSpec columnsPairSpec) {
        return new TableComparisonGroupingColumnPairModel() {{
            setComparedTableColumnName(columnsPairSpec.getComparedTableColumnName());
            setReferenceTableColumnName(columnsPairSpec.getReferenceTableColumnName());
        }};
    }

    /**
     * Creates a grouping column pair specification object from this model.
     * @return Grouping column pair specification.
     */
    public TableComparisonGroupingColumnsPairSpec createColumnsPairSpec() {
        TableComparisonGroupingColumnsPairSpec result = new TableComparisonGroupingColumnsPairSpec();
        result.setComparedTableColumnName(this.getComparedTableColumnName());
        result.setReferenceTableColumnName(this.getReferenceTableColumnName());
        return result;
    }
}

File: distribution/python/dqops/client/models/table_comparison_grouping_column_pair_model.py
from typing import Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

T = TypeVar("T", bound="TableComparisonGroupingColumnPairModel")


@_attrs_define
class TableComparisonGroupingColumnPairModel:
    """Model that identifies a pair of column names used for grouping the data on both the compared table and the reference
    table. The groups are then matched (joined) by DQOps to compare aggregated results.

        Attributes:
            compared_table_column_name (Union[Unset, str]): The name of the column on the compared table (the parent table)
                that is used in the GROUP BY clause to group rows before compared aggregates (row counts, sums, etc.) are
                calculated. This column is also used to join (match) results to the reference table.
            reference_table_column_name (Union[Unset, str]): The name of the column on the reference table (the source of
                truth) that is used in the GROUP BY clause to group rows before compared aggregates (row counts, sums, etc.) are
                calculated. This column is also used to join (match) results to the compared table.
    """

    compared_table_column_name: Union[Unset, str] = UNSET
    reference_table_column_name: Union[Unset, str] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        compared_table_column_name = self.compared_table_column_name
        reference_table_column_name = self.reference_table_column_name

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if compared_table_column_name is not UNSET:
            field_dict["compared_table_column_name"] = compared_table_column_name
        if reference_table_column_name is not UNSET:
            field_dict["reference_table_column_name"] = reference_table_column_name

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        compared_table_column_name = d.pop("compared_table_column_name", UNSET)

        reference_table_column_name = d.pop("reference_table_column_name", UNSET)

        table_comparison_grouping_column_pair_model = cls(
            compared_table_column_name=compared_table_column_name,
            reference_table_column_name=reference_table_column_name,
        )

        table_comparison_grouping_column_pair_model.additional_properties = d
        return table_comparison_grouping_column_pair_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
